lib/discover_endpoint.js

```diff
@@ -0,0 +1,340 @@
+var AWS = require('./core');
+var util = require('./util');
+var ENDPOINT_OPERATION_MAX_RETRIES = 60;
+var endpointDiscoveryEnabledEnv = 'AWS_ENABLE_ENDPOINT_DISCOVERY';
+
+/**
+ * Generate key to index the endpoints in the cache
+ * @return [map<String,String>] object with keys to index endpoints.
+ * @api private
+ */
+function getCacheKey(request) {
+  var service = request.service;
+  var api = service.api || {};
+  var operations = api.operations;
+  var identifiers = {};
+  if (operations[request.operation] && operations[request.operation].name) {
+    identifiers.operation = operations[request.operation].name;
+  }
+  if (service.config.region) {
+    identifiers.region = service.config.region;
+  }
+  if (api.serviceId) {
+    identifiers.serviceId = api.serviceId
+  }
+  if (service.config.credentials.accessKeyId) {
+    identifiers.accessKeyId = service.config.credentials.accessKeyId
+  }
+  return identifiers;
+}
+
+/**
+ * @api private
+ */
+function marshallCustomIdentifiersHelper(result, params, shape) {
+  if (!shape || params === undefined || params === null) return;
+  if (shape.type === 'structure' && shape.required && shape.required.length > 0) {
+    util.arrayEach(shape.required, function(name) {
+      var memberShape = shape.members[name];
+      if (memberShape.endpointDiscoveryId === true) {
+        var locationName = memberShape.isLocationName ? memberShape.name : name;
+        result[locationName] = String(params[name]);
+      } else {
```

Comment:
What would the model have to look like for this else statement to find an `endpointDiscoveryId`?

Reply:
It will look for the `endpointDiscoveryId` in depth 2. According to the design, `endpointDiscoveryId` is also possible to exist as a required string member inside a top level required structure member.